### PR TITLE
Migrate text layers to nodes

### LINKED ISSUE #768: Replace legacy layer types with nodes
**Task: Switch LayerDataType struct values to nodes**

```rs
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
/// Represents different types of layers.
pub enum LayerDataType {
	/// A layer that wraps a [FolderLayer] struct.
	Folder(FolderLayer),
	/// A layer that wraps a [ShapeLayer] struct.
	Shape(ShapeLayer),
	/// A layer that wraps a [TextLayer] struct.
	Text(TextLayer),
	/// A layer that wraps an [ImageLayer] struct.
	Image(ImageLayer),
        Imaginate(ImaginateLayer),
}
```
~~`FolderLayer`~~, `ShapeLayer`, `TextLayer`, and `ImageLayer` live in `/graphene`. Each of these structs (saving `FolderLayer` for another issue) should be converted into nodes using the new node system in `/node-graph`.

---

*Questions for @TrueDoctor:*
- *How does this approach relate to the separation between "document graph" and "proto graph"?*
- *Can we continue to reference the other data structures defined in `/graphene` which are used by those three (four) structs once they are ported to nodes in `/node-graph`?*
- *Where do these new vector-editing-specific nodes go in the node graph folder structure?*
- *Any other things to be aware of?*

### LINKED ISSUE #998: Port Text layers to nodes

### LINKED ISSUE #1163: Stop vector layers disappearing whilst dragging
Currently, vector layers disappear whilst dragging them, causing some confusion. This is not the desired behaviour, they should remain visible whilst dragging or transforming with G/R/S.

### PR DESCRIPTION
**Pull request by 0HyperCube**
Closes #998, closes #1163, closes #768

- [x] Node impl
- [x] Properties
- [x] Preview
- [x] Font cache to graph
- [x] Fix transforms
- [x] Enter text tool on double click
- [x] Fix error when using path tool
- [x] Fix text sometimes not being rendered
- [ ] Editing the text in the properties panel causes the position of the text in the document to shift (we need to update the position when the text changes, as the position is relative to the centre of the text)

### CODE REVIEW

**Comment by Keavon** [comment]
A bug: when reloading the page (to auto-save and load), the text shrinks to 1x1 pixel.

File: node-graph/gcore/src/raster.rs [hunk 1]
```diff
@@ -532,7 +532,7 @@ fn dimensions_node<_P>(input: ImageSlice<'input, _P>) -> (u32, u32) {
 #[cfg(feature = "alloc")]
 pub use image::{CollectNode, Image, ImageFrame, ImageRefNode, MapImageSliceNode};
 #[cfg(feature = "alloc")]
-mod image {
+pub(crate) mod image {
```
Comment (TrueDoctor) [inline node-graph/gcore/src/raster.rs:535]: We should probably move this to another file, but we can do that in another pr if you prefer

File: node-graph/gcore/src/text/font_cache.rs [hunk 2]
```diff
@@ -64,3 +65,15 @@ impl FontCache {
 		self.preview_urls.get(font)
 	}
 }
+
+impl core::hash::Hash for FontCache {
+	fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
+		self.preview_urls.len().hash(state);
```
[the remark this answers is not in the document]
Comment (0HyperCube) [reply node-graph/gcore/src/text/font_cache.rs:71]: Probably true, but I was running into some issues with hash collisions leading to some strange and difficult to debug results.

File: node-graph/gcore/src/text.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	#[inline]
+	fn eval(&'input self, editor: EditorApi<'a>) -> Self::Output {
+		let text = self.text.eval(());
+		let font = self.font_name.eval(());
+		let font_size = self.font_size.eval(());
+
+		let buzz_face = editor.font_cache.and_then(|cache| cache.get(&font)).map(|data| load_face(data));
+		crate::vector::VectorData::from_subpaths(to_path(&text, buzz_face, font_size, None))
+	}
+}
```
[the remark this answers is not in the document]
Comment (0HyperCube) [reply node-graph/gcore/src/text.rs:32]: Yes, it runs into the error:
```sh
error: implementation of `graphene_core::Node` is not general enough
   --> node-graph/interpreted-executor/src/node_registry.rs:52:5
    |
52  |                 Box::pin(any)
    |                 ^^^^^^^^^^^^^ implementation of `graphene_core::Node` is not general enough
...
512 |         register_node!(graphene_core::text::TextGenerator<_, _, _>, input: graphene_core::EditorApi, params: [String, graphene_core::text::Font, f64]),
    |         ---------------------------------------------------------------------------------------------------------------------------------------------- in this macro invocation
    |
    = note: `TextGenerator<ClonedNode<std::string::String>, ClonedNode<graphene_core::text::Font>, ClonedNode<f64>>` must implement `graphene_core::Node<'0, graphene_core::EditorApi<'_>>`, for any lifetime `'0`...
    = note: ...but it actually implements `graphene_core::Node<'1, graphene_core::EditorApi<'1>>`, for some specific lifetime `'1`
    = note: this error originates in the macro `register_node` (in Nightly builds, run with -Z macro-backtrace for more info)
```
Which was somewhat confusing as I would have assumed that it could be the 'input lifetime.

File: editor/src/messages/portfolio/document/node_graph/node_graph_message_handler/document_node_types.rs [hunk 4]
```diff
@@ -1098,3 +1114,43 @@ pub fn new_vector_network(subpaths: Vec<bezier_rs::Subpath<uuid::ManipulatorGrou
...
+			fill.to_document_node_default_inputs([Some(NodeInput::node(1, 0))], next_pos()),
+			stroke.to_document_node_default_inputs([Some(NodeInput::node(2, 0))], next_pos()),
+			output.to_document_node_default_inputs([Some(NodeInput::node(3, 0))], next_pos()),
+		]
+		.into_iter()
+		.enumerate()
+		.map(|(id, node)| (id as NodeId, node))
+		.collect(),
+		..Default::default()
+	}
```
[the remark this answers is not in the document]
Comment (0HyperCube) [reply editor/src/messages/portfolio/document/node_graph/node_graph_message_handler/document_node_types.rs:1155]: Sure, I hadn't seen that.

File: editor/src/messages/portfolio/document/document_message_handler.rs [hunk 5]
```diff
@@ -1056,36 +1040,39 @@ impl DocumentMessageHandler {
...
 		// Find the primary input type of the node graph
-		let primary_input_type = node_network.input_types().next().clone();
-		let response = match primary_input_type {
-			// Only calclate the frame if the primary input is an image
-			Some(ty) if ty == concrete!(ImageFrame<Color>) => {
-				// Calculate the size of the region to be exported
-				let old_transforms = self.remove_document_transform();
-				let transform = self.document_legacy.multiply_transforms(&layer_path).unwrap();
-				let size = DVec2::new(transform.transform_vector2(DVec2::new(1., 0.)).length(), transform.transform_vector2(DVec2::new(0., 1.)).length());
+		let input_frame = node_network.nodes.iter().find(|(_, node)| node.name == "Input Frame");
```
[the remark this answers is not in the document]
Comment (TrueDoctor) [reply editor/src/messages/portfolio/document/document_message_handler.rs:1043]: Ah true, I forgot about that. I guess this will only really get resolved once we transition to having a screenshot node. Do you mind adding a TODO comment explaining that?

File: node-graph/gcore/src/text/font_cache.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (TrueDoctor) [reply node-graph/gcore/src/text/font_cache.rs:71]: Oh, that is fascinating, definetly good to know

File: editor/src/messages/portfolio/document/document_message_handler.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Keavon) [reply editor/src/messages/portfolio/document/document_message_handler.rs:1043]: *(just a note: we may or may not end up refactoring the screenshot node, or we might just jump straight into full nodes, depending on timing*)

[the remark this answers is not in the document]
Comment (TrueDoctor) [reply editor/src/messages/portfolio/document/document_message_handler.rs:1043]: the screenshot node would be the async fully node version. (You could also call it rasterize if you want)

[the remark this answers is not in the document]
Comment (Keavon) [reply editor/src/messages/portfolio/document/document_message_handler.rs:1043]: Yeah, I'm just talking about how we might end up being very close to the full node system which would obviate the screenshot node entirely. The screenshot node (regardless of its name) is just a stopgap until then.

File: node-graph/gcore/src/text.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (TrueDoctor) [reply node-graph/gcore/src/text.rs:32]: That's unfortunate, ran into similar issues with the `CloneNode`. Did you attempt to put the lifetime into the fn generics?
```rust
#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Default)]
pub struct CloneNode<O> {
	_o: PhantomData<O>,
}
#[node_macro::node_fn(CloneNode<_O>)]
fn clone<'n: 'input, _O: Clone>(input: &'n _O) -> _O {
	input.clone()
}

```

**Review by Keavon — APPROVED** [review]
